docs/architecture/adr-050-sign-mode-textual.md

```diff
@@ -129,6 +130,30 @@ Msg (<int>/<int>): <string>           // E.g. Msg (1/2): bank v1beta1 send coins
 End of transaction messages
 ```
 
+### 9. Custom `Msg`-Renderers
+
+Application developers may choose to not follow default value renderers' output for their own `Msg`s. In this case, they can implement their own custom `Msg`-renderer. This is similar to [EIP4430](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4430.md), where the smart contract developer chooses the description string to be shown to the end user.
+
+This is done by setting the `cosmos.msg.v1.textual.custom_renderer` Protobuf option to `true`. This option MUST be set on a `sdk.Msg`, i.e. on a Protobuf message whose name starts with `Msg`.
+
+```proto
+message MsgFooBar {
+  option (cosmos.msg.v1.textual.custom_renderer) = true;
+
+  // proto fields
+}
+```
+
+When this option is set to true on a `Msg`, the implementation CANNOT use the default value renderers to render the `Msg`. Instead, it MUST check that a custom `Msg`-renderer is implemented for `MsgFooBar` (e.g. registered in a registry), or else the implementation MUST throw an error or panic.
+
+Moreover, the implementation must provide 2 functions: one for formatting from Protobuf to string, and one for parsing string to Protobuf. These 2 functions are provided by the application devloper. To satisfy point #1, these 2 functions MUST be bijective with each other as inverse.
+
+Bijectivity of custom `Msg`-renderers SHOULD be checked with best effort by the SDK on node startup. To achieve this, we plan to use a similar technique to property testing, whereby a set of random inputs are generated, and passed into the relevant `Msg`. We then test that the composition of the two custom functions, formatting and parsing, result in the original Protobuf `Msg`:
```

Comment:
I'd recommend that you either trust the custom rendering (and check bijectivity through provided tests) or don't (and check every time at runtime). I don't see the additional safety of checking at node startup time vs checking the software release - nothing should be different at node startup time.

Reply:
Agree.

Reply:
The question here is: Should the Cosmos SDK check that the app dev's custom rendering is bijective?

> you either trust the custom rendering (and check bijectivity through provided tests)

How would the SDK check bijectivity through provided tests (which I suppose will live in `_test.go` files in the app dev's repo)?

Or the SDK checks nothing. But then this is really a footgun.

> or don't (and check every time at runtime)

The way I see it, the SDK will expose a `textual.RegisterCustomMsgRenderer()` to app chains. Probably somewhere in app.go, app devs will call it with their custom implementation. I was thinking to check bijectivity upon this function call, which happens at node startup.

We could also check bijectivity at runtime, i.e. when verifying a tx signature which uses TEXTUAL. But this will slow down performance, with no added benefit vs at node startup.

Reply:
Yes, a comprehensive suite would be part of the app repo, and should be there regardless of where else checks might run.

The problem with a startup check is for applications like ours where the transformation depends on context, e.g. having a set of per-account petnames for addresses. It's easy to set up such context in a test and easy to check the real use case at runtime, but less straightforward and less representative of real use to check on startup. Do you make a real account on your mainnet just for testing? Now you've got to update this state if the tests change.

Because of these difficulties, I think a startup test would tend to be trivial, and a trivial test provides little protection compared to a comprehensive test suite, which would be inappropriate to run on startup.

Having the inverse mapping be part of the "Renderer" interface is a strong signal to developers for how this feature should be designed and tested. At some point you need to trust the developer, otherwise you won't have any powerful features.

Reply:
That's a good point. With state reads, the node startup check would indeed be trivial. I rephrased the ADR to say that the SDK won't check bijectivity, and it's the responsibility of the app developer to check it.